TN5 withdrawal devnet test

philippecamacho opened the pull request:
Closes https://app.asana.com/1/1208976916964769/project/1209423958092927/task/1210592590143353?focus=true


### This PR:
Implements a test that checks it is possible for a user to withdraw funds from L2 to L1.

### This PR does not:
Make any change to the integration code itself.

### Key places to review:
Focus on `withdraw_test.go`.

### How to test this PR:  
```
just devnet-withdrawal-test
```

philippecamacho committed:
Basic test running in devnet environment.

philippecamacho committed:
Fetch L1 balance

philippecamacho committed:
Check Alice's balance on L1

philippecamacho committed:
Trying to publish a withdraw proof, yet the proposer does not submit the game apparently.

philippecamacho committed:
Proposer manages to publish the game.

philippecamacho committed:
More logging

philippecamacho committed:
More debugging messages.

philippecamacho committed:
Add function to get the right addresses for the contracts.

philippecamacho committed:
Withdrawal proof verified successfully.

philippecamacho committed:
Attempt to transfer funds to user.

philippecamacho committed:
Comment code that still fails.

philippecamacho committed:
Merge justfile

philippecamacho committed:
Remove comments.

philippecamacho committed:
Remove redundant imports

philippecamacho committed:
Address possible rebase issues.

philippecamacho committed:
Do not delete the op-deployer binary.ç

philippecamacho committed:
Run smoke test. It passes!

philippecamacho committed:
Run all the devnet tests locally.

philippecamacho committed:
Don't run the batcher restart test on CI for now.

philippecamacho committed:
Test passed at least once.

philippecamacho committed:
Trying to improve docker shutdown.

philippecamacho committed:
Add smoke test for devnet

philippecamacho committed:
Challenge game test passing locally.

philippecamacho committed:
Run smoke test in CI.

philippecamacho committed:
Run more tests in CI

philippecamacho committed:
Adding batcher restart test to CI.

philippecamacho committed:
Fix Conflicts:

philippecamacho committed:
YConflict fixes

philippecamacho committed:
Able to fetch the withdrawl delay from the optimism portal contract.

philippecamacho committed:
Fix github actions.

philippecamacho committed:
Remove simple hash computation.

philippecamacho committed:
Clean up.

philippecamacho committed:
Cleaner hash computation.

philippecamacho committed:
More cleanup

philippecamacho committed:
Add debug API to L1 RPC.

philippecamacho committed:
Finalize game.

philippecamacho committed:
Cleanup

Ayiga commented on errors.txt:
We probably don't want to commit this file to the repo

philippecamacho replied:
Yes, sorry for that. I have added "errors.txt" to .gitgnore.

philippecamacho committed:
Remove getOPAddresses function

philippecamacho committed:
Test passing until last step

philippecamacho committed:
Useful debugging logs.

philippecamacho committed:
Trying out things for dispute resolution.

philippecamacho committed:
Cleanup test

philippecamacho committed:
Refactor the test

philippecamacho committed:
Implementation of L1 deposit function.

philippecamacho committed:
All tests passing...

philippecamacho committed:
Increase withdraw amount

philippecamacho committed:
Check Alice L1 balance increases correctly.

philippecamacho committed:
Revert changes to OptimismPortal2.sol

philippecamacho committed:
Revert "Revert changes to OptimismPortal2.sol"

This reverts commit e19a6ddd05b1216be67b8d95b209a4a12b57b5e9.

philippecamacho committed:
Readd OptimismPortal_InvalidProofTimestamp() check.

philippecamacho committed:
resolveGame is not needed.

philippecamacho committed:
Readd OptimismPortal_InvalidRootClaim. Test fails.

philippecamacho committed:
Only one revert commented. Test passing.revert OptimismPortal_InvalidRootClaim()

philippecamacho committed:
Revert last change in OptimismPortal. Managed to configure DisputeMaxClockDuration. Test still fails though.

philippecamacho committed:
Test passing

philippecamacho committed:
Fix golint errors

philippecamacho committed:
Fix golint errors.

philippecamacho committed:
Cleanup

philippecamacho committed:
Remove getWithtdrawDelay function

philippecamacho committed:
More simplification.

Ayiga committed:
Replace gameCtx timeout with shorter value and retries

In the `waitForGameToBePublished` function a `gameCtx` is created with a timeout of
`5 minutes`.  This `gameCtx` is passed to `wait.ForGamePublished` which
immediately adds its own timeout of `2 minutes` which will trigger before the
`5 minute` timeout.  To get around this, we opt for a shorter timeout of `1 minute`,
and we retry up to 5 times to account for the total `5 minute` duration while also
minimizing the total time needed to wait.

philippecamacho committed:
Simplification of the code.

philippecamacho committed:
Reduce a bit the time for running the test.

shenkeyao commented on .gitignore:
```diff
@@ -66,3 +66,7 @@ gha-creds-*.json
 *.pem
+
+
+# Local logs
+errors.txt
```
Curious about this--is it a file you added locally? Maybe we don't need it here if it's not a file that may be generated on everyone's machine.

philippecamacho replied:
Yes, this is because I run `just devnet-withdrawal-test > errors.txt` locally. I [accidentally](https://github.com/EspressoSystems/optimism-espresso-integration/pull/226#discussion_r2372405923) added the errors.txt file to the repository earlier.

shenkeyao replied:
Got it! I think we should only add a file to `gitignore` if it's generated by a script that everyone uses, but we don't want to store it in GitHub. For example, if we have a `test.sh` script which always outputs errors to `test-errors.txt` to help investigations, then we probably want to add `test-errors.txt` to `gitignore`.

But in this case, you manually specified `> errors.txt` to create the file, which no one else will do, so I feel like it doesn't belong in `gitignore`. Otherwise, it will end up being a long list of everyone's customized files.

Sorry for nitpicking! I just feel like it's better if we just don't commit the temporary file in this case. But let me know if this doesn't make sense.

philippecamacho replied:
Fair enough, the change in .gitignore has been removed in 01f506ed3489c3b53cef0a7af424774827303150.

shenkeyao commented on espresso/devnet-tests/withdraw_test.go:
```diff
@@ -0,0 +1,332 @@
+	require.NoError(t, err)
+
+	// Verify balance change
+	_, err = wait.ForBalanceChange(ctx, d.L1, userAddress, withdrawalAmount)
```
The last parameter of `ForBalanceChange` is the initial balance, so shouldn't it be `balanceBefore`?

philippecamacho replied:
You are right! Fixed in [6e547a3](https://github.com/EspressoSystems/optimism-espresso-integration/pull/226/commits/6e547a378a06de271ba8c98a2f41000d95aa5200).

philippecamacho committed:
Fix parameter passed to wait.ForBalanceChange.

shenkeyao reviewed (approved):
Left a long but minor comment. Otherwise, this looks good!

philippecamacho committed:
Remove mention of errors.txt in .gitignore.
